src/user.rs

```diff
@@ -50,73 +52,66 @@ pub(crate) async fn user_main(
-                } else {
-                    0
-                };
-
-                // Counter to track how long we've slept, waking regularly to check for messages.
-                let mut slept: usize = 0;
-
-                // Wake every second to check if the parent thread has told us to exit.
-                let mut in_sleep_loop = true;
-                // Track the time slept for Coordinated Omission Mitigation.
-                let sleep_timer = time::Instant::now();
-                while in_sleep_loop {
-                    let mut message = thread_receiver.try_recv();
-                    while message.is_ok() {
-                        match message.unwrap() {
-                            // Time to exit, break out of launch_tasks loop.
-                            GooseUserCommand::Exit => {
-                                break 'launch_tasks;
-                            }
-                            command => {
-                                debug!("ignoring unexpected GooseUserCommand: {:?}", command);
-                            }
-                        }
-                        message = thread_receiver.try_recv();
+        let mut task_iter = thread_task_set.weighted_tasks.iter().cycle();
+        let next_task = Fuse::terminated();
+        pin_mut!(next_task);
+
+        let task_wait = match thread_task_set.task_wait.take() {
+            Some((min, max)) if min == max => min,
+            Some((min, max)) => Duration::from_millis(
+                rand::thread_rng().gen_range(min.as_millis()..max.as_millis()) as u64,
+            ),
+            None => Duration::from_millis(0),
+        };
+
+        next_task.set(tokio::time::sleep(Duration::from_secs(0)).fuse());
+        loop {
+            select! {
+                _ = next_task => {
+                    let (thread_task_index, thread_task_name) = task_iter.next().unwrap();
+                    if *thread_task_index == 0 {
+                        // Tracks the time it takes to loop through all GooseTasks when Coordinated Omission
+                        // Mitigation is enabled.
+                        thread_user.update_request_cadence(thread_number).await;
                     }
-                    if thread_user.max_wait > 0 {
-                        let sleep_duration = time::Duration::from_secs(1);
-                        debug!(
-                            "user {} from {} sleeping {:?} second...",
-                            thread_number, thread_task_set.name, sleep_duration
-                        );
-                        tokio::time::sleep(sleep_duration).await;
-                        slept += 1;
-                        if slept > wait_time {
-                            in_sleep_loop = false;
-                        }
+
+                    // Determine which task we're going to run next.
+                    let task = &thread_task_set.tasks[*thread_task_index];
+                    let function = &task.function;
```

Comment:
This comment doesn't really make sense anymore, perhaps combine these two lines of code and rework the comment as follows:

```rust
                    // Get a reference to the task function we're going to invoke next.
                    let function = &thread_task_set.tasks[*thread_task_index].function;
```